### DISCUSSION TITLE
Error handling for applications: how to return a public message error instead of all the chain of errors and tracing that chain at the same time?

### DISCUSSION THREAD

**Discussion by frederikhors** [async-graphql/async-graphql#1089] [Q&A]
### PROLOGUE

Dear guys, I have hundreds of resolvers and for each resolver I would like to trace all the possible errors.

In each method of my app I'm using `anyhow::{Error}`.

Right now I have code similar to this for each resolver:

```rust
#[Object]
impl MutationRoot {
    async fn player_create(&self, ctx: &Context<'_>, input: PlayerInput) -> Result<Player> {
        let services = ctx.data_unchecked::<Services>();

        let player = services
            .player_create(input)
            .await?;

        Ok(player)
    }
}
```

So I thought about using the below code (_note the added line with `.map_err()`_):

```rust
#[Object]
impl MutationRoot {
    async fn player_create(&self, ctx: &Context<'_>, input: PlayerInput) -> Result<Player> {
        let services = ctx.data_unchecked::<Services>();

        let player = services
            .player_create(input)
            .await
            .map_err(errorify)?;

        Ok(player)
    }
}

fn errorify(err: anyhow::Error) -> async_graphql::Error {
    tracing::error!("{:?}", err);

    err.into()
}
```

Now the error is traced along with all the error chain:

```
ERROR example::main:10: I'm the error number 4

Caused by:
    0: I'm the error number 3
    1: I'm the error number 2
    2: I'm the error number 1
    in example::async_graphql
```

### QUESTION 1

Is there a way to avoid the `.map_err()` on each resolver?

I would like to use the `?` alone.

Should I use a custom error?

Can we have a global "hook"/callback/fn to call on each error?

### QUESTION 2

As you can see above the chain of the error is the inverse.

In my graphql response I'm getting as `message` the "I'm the error number 4" but I need to get the "I'm the error number 2" instead.

The error chain is created using `anyhow` like this:

- main.rs: returns error with `.with_context(|| "I'm the error number 4")?`
- - call `fn player_create()` in `graphql.rs`: returns with `.with_context(|| "I'm the error number 3")?`
- - - call `fn new_player()` in `domain.rs`: returns with `.with_context(|| "I'm the error number 2")?`
- - - - call `fn save_player()` in `database.rs`: returns with `.with_context(|| "I'm the error number 1")?`

**How can I accomplish this?**

I'm really new to Rust. I come from Golang where I was using a struct like:

```go
type Error struct {
	error          error
	public_message string
}
```

chaining it easily with:

```go
return fmt.Errorf("this function is called but the error was: %w", previousError)
```

**How to do it in Rust?**

Do I necessarily have to use `anyhow`?

Can you point me to a good handling error tutorial/book for applications?

Thank you very much.

**Comment by sunli829** [comment]
> QUESTION 1

You can use this [extension](https://docs.rs/async-graphql/4.0.14/async_graphql/extensions/struct.Tracing.html), or implement a extension by youself.

> QUESTION 2

```rust
async_graphql::Error::from(format!("this function is called but the error was: {}", previous_error))
```

This basically does the same thing as your Go code.

**Reply by frederikhors** [reply]
> You can use this [extension](https://docs.rs/async-graphql/4.0.14/async_graphql/extensions/struct.Tracing.html), or implement a extension by youself.

I'm already using it. I created a reproduction for you here: https://github.com/frederikhors/iss-async-graphql-error-handling.

You can read the [README.md](https://github.com/frederikhors/iss-async-graphql-error-handling/blob/main/README.md) to understand the issue I'm having.

_Sorry to bother you: I will write a PR for newbies like me to repay you for your precious help._

**Reply by frederikhors** [reply]
I created https://github.com/async-graphql/async-graphql/pull/1091.
